### DISCUSSION TITLE
Can the Checks API be used to trigger two runs when two pushed branches point at the same commit?

### DISCUSSION THREAD

**Discussion by foolip** [community/community#24475] [Apps, API and Webhooks]
<p>In <a href="https://github.com/web-platform-tests/wpt" rel="nofollow noopener">https://github.com/web-platform-tests/wpt</a> we do some amount of testing for each push to the master branch, but also have epochs/daily and epochs/weekly branches trailing the master branch which we’ve used to trigger slightly different jobs. (The project is for testing browsers, and we test stable/beta versions less often than dev versions of the browser.)</p>
<p>This setup is working with Mozilla’s Taskcluster CI system, which is currently using the old Status API integration. However, I’m unable to do the equivalent thing with Azure Pipelines which is using the Checks API integration. <a href="https://github.com/web-platform-tests/wpt/issues/14836" rel="nofollow noopener">https://github.com/web-platform-tests/wpt/issues/14836</a> has the details of what I have attempted.</p>
<p>It appears that because checks are triggered for a commit sha and not a symbolic ref (branch name), that if the push to master has already triggered checks, later pushing another branch pointing to the same commit will not create a new check suite or cause any check runs to rerun.</p>
<p>I would like to confirm that my understanding of the Checks API is correct, and if this is intentional. Are there any known workarounds? It does appear that some uses of <a href="https://docs.microsoft.com/en-us/azure/devops/pipelines/build/triggers" rel="nofollow noopener">https://docs.microsoft.com/en-us/azure/devops/pipelines/build/triggers</a> are fundamentally incompatible with the Checks API model, but I’d like to confirm this before reporting to the Azure Pipelines team.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>Yes, Checks run against an individual commit. To my knowledge, they do not take into account the number of symbolic refs that may or may not refer to that individual commit because that will change over time. However, since that commit is immutable, so long as the Checks are looking at the contents of the repository and not anything else, new runs of the Checks should create the same result. However, a Check Suite can be manually rerun through the UI or <a href="https://developer.github.com/v3/checks/suites/#rerequest-check-suite" rel="nofollow noopener">requested through the API</a>. You could potentially create <a href="https://github.com/features/actions" rel="nofollow noopener">an Action</a> or a <a href="https://probot.github.io" rel="nofollow noopener">Probot App</a> to create the workflow you desire by requesting a rerun of a Check Suite when a new branch is pushed and the Suite is already completed?</p>

**Comment by foolip** [comment]
<p>We have a workaround now where we don’t rely on Checks at all but instead have a scheduled build in Azure Pipelines.</p>
<p>However, I noticed while playing with GitHub Actions that the problem doesn’t seem to be intrinsic to Checks. In <a href="https://github.com/foolip/wpt/runs/56065527" rel="nofollow noopener">https://github.com/foolip/wpt/runs/56065527</a>, I pushed the same ref to two different branches, and that resulted in two different check suites which ran at different times. Screenshot:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/5c63496ec5c5a543737456361596584e61594d49.png" data-download-href="/uploads/short-url/dbiEckszSsn8p96Wk7VKSV7kOOR.png?dl=1" title="Screenshot from 2019-01-29 13-03-52.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/498917/181084825-f4f8ae97-6fd7-4b57-8b06-efc19a3f3abf.png" alt="Screenshot from 2019-01-29 13-03-52.png" data-base62-sha1="dbiEckszSsn8p96Wk7VKSV7kOOR" width="690" height="463" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/5/5c63496ec5c5a543737456361596584e61594d49_2_690x463.png, https://aws1.discourse-cdn.com/github/optimized/2X/5/5c63496ec5c5a543737456361596584e61594d49_2_1035x694.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/5/5c63496ec5c5a543737456361596584e61594d49.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/5c63496ec5c5a543737456361596584e61594d49_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot from 2019-01-29 13-03-52.png</span><span class="informations">1266×851 106 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Is GitHub Actions using internal APIs available only to GitHub, or could any CI system achieve the same?</p>

**Comment by lee-dohm** [comment]
<p>According to <a href="https://github.com/foolip/wpt/blob/master/.github/main.workflow" rel="nofollow noopener">the main.workflow</a> that I’m looking at in your sample repository, the workflow <a href="https://github.com/foolip/wpt/blob/master/.github/main.workflow?short_path=7e53c47#L2" rel="nofollow noopener">triggers on <code>push</code></a>. Checks works slightly differently in that it creates a Check Suite object for a commit when it is pushed and then notifies the GitHub App via a <code>check_suite</code> event, rather than a <code>push</code> event.</p>
<p>I hope that helps!</p>

**Comment by foolip** [comment]
<p>That is what I suspected. Perhaps a GitHub app could also react to push events and manually create check suites to get the same behavior?</p>

**Comment by lee-dohm** [comment]
<p>It’s all a matter of subscribing to the events you want when you create or update your GitHub App:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/45d42e4c8d7ffff98ffe227d85db0ecf11e193a4.png" data-download-href="/uploads/short-url/9XJxx16Z17KJyMLEL26uQ7GEYWE.png?dl=1" title="Screen Shot 2019-01-30 at 8.00.52 AM.png"><img src="https://user-images.githubusercontent.com/1038121/181084830-604cf1a6-0f6f-4a1b-8d26-d3a825018296.png" alt="Screen Shot 2019-01-30 at 8.00.52 AM.png" data-base62-sha1="9XJxx16Z17KJyMLEL26uQ7GEYWE" width="472" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/4/45d42e4c8d7ffff98ffe227d85db0ecf11e193a4_2_472x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/4/45d42e4c8d7ffff98ffe227d85db0ecf11e193a4_2_708x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/4/45d42e4c8d7ffff98ffe227d85db0ecf11e193a4_2_944x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/45d42e4c8d7ffff98ffe227d85db0ecf11e193a4_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-01-30 at 8.00.52 AM.png</span><span class="informations">1738×1840 351 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by foolip** [comment]
<p>Alright, thanks for your help. I’ll point developers at Azure Pipelines and Taskcluster at this thread when suggested that they match  GitHub Actions’s behavior on this.</p>
